### PR TITLE
Search is a team work

### PR DESCRIPTION
**Pull request by mauroservienti**

### CODE REVIEW

File: _posts/2019-05-22-search-is-a-team-work.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+Let's use the Hotel Reservations scenario as a way to identify some requirements:
+
+* Users need to be able to search hotels by location, .e.g. a city.
+* The search criteria can contain a check-in date, and/or a check-out date to filter out unavailable hotels.
+* Results can be further filtered by:
+  * amenities, such as breakfast included, or a spa. 
+  * stars
+  * guests reviews and ratings
+
+By looking at the above requirements and doing a very rough, and highly unsuggested, boundaries analysis, we could identify that:
```
Comment (tmasternak) [inline _posts/2019-05-22-search-is-a-team-work.md]:
``` suggestion
One can safely assume that the search attributes won't belong to a single service. We won't get into the service boundaries identification here though and for the sake of simplicity assume that we are dealing with the following decomposition:
```

File: _posts/2019-05-22-search-is-a-team-work.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+* Registry owns:
+  * the hotel location
+  * amenities
+  * stars
+* Reservations owns availability
+* Marketing owns guests reviews and ratings
+
+## A naïf design
+
+If we look at the above requirements, as they are listed, in theory we could design a distributed search process like the following (by applying some [ViewModel Composition techniques](https://milestone.topics.it/categories/view-model-composition)):
```
Comment (tmasternak) [inline _posts/2019-05-22-search-is-a-team-work.md]:
```suggestion
If we look at the above requirements, as they are listed, in theory, we could design a distributed search process by applying [ViewModel Composition techniques](https://milestone.topics.it/categories/view-model-composition) in the following way:
```

File: _posts/2019-05-22-search-is-a-team-work.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+* Registry will handle the first part of the query returning the list of hotel IDs matching the `location` criteria
+* Returned IDs are then handover to Reservations that filters out, based on selected dates, IDs matching hotels not available
```
Comment (tmasternak) [inline _posts/2019-05-22-search-is-a-team-work.md]:
``` suggestion
* Returned IDs are then handed over to Reservations that filters out IDs based on availability in the selected search period
```

File: _posts/2019-05-22-search-is-a-team-work.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+* Finally IDs are passed to Marketing, and again Registry, to execute a [Faceted search](https://en.wikipedia.org/wiki/Faceted_search) and build facets results.
+
+It works, still is very inefficient. It would be like if the engineers from the header image of this article were to move the race car to different locations to change tyres or to refuel it. Very inefficient.
```
Comment (tmasternak) [inline _posts/2019-05-22-search-is-a-team-work.md]: Should this paragraph state explicitly that too much data is being moved between storage and composition process or even between processes via remoting? - to drive the point home?

Comment (mauroservienti) [reply _posts/2019-05-22-search-is-a-team-work.md]: Yes, I’ll reword it a bit.

Comment (mauroservienti) [reply _posts/2019-05-22-search-is-a-team-work.md]: Reworded. I’ll go ahead and merge this. We can always improve later, and there is still an editorial review pending.

**Pull request merged**